text/0255-sensitive-data-redaction.md

```diff
@@ -0,0 +1,397 @@
+- methods for consumers of the OpenTelemetry API to enrich collected Attributes
+  with sensitivity information and hooks to apply different levels of redaction.
+- A redactor implements the logic to apply redaction and that owns predefined helpers
+  for redaction in the SDK (URLParams filtering, Zeroing IPs, etc.).
+
+The following limitations apply:
+
+- Although it is possible that sensitive data is contained in any data generated
+  by OpenTelemetry (e.g. the span name, or the instrumentation scope name could
+  contain sensitive data), only the following is treated in the scope of this PR:
+  [Attributes](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/common/README.md#attribute)
+  and payload fields like [log bodies](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/data-model.md#field-body).
+
+### Configuration-driven redaction
+
+As stated in the motivation the responsibility for applying redaction lays with
+the operator of an application, since they are the only ones knowing the specific
+requirements for their environment. Because of that this proposal suggests that
+OpenTelemetry should provide a solution to end users that allows them to provide
+a configuration for their redaction requirements. This configuration consists of
+two pieces:
+
+* Redaction requirement "configuration profiles", that end user use to express
+  their overall redaction requirements.
+* Fine grained configuration that can be used to express requirements that are
+  unique to their environment.
+
+**Example**: An end user may configure an application with "stricter" requirements
+ and some custom requirements like the following:
+
+* They provide a profile name `STRICTER` via an environment variable, e.g.
+  OTEL_SENSITIVE_DATA_PROFILE="STRICTER"
+* They add custom requirements in a
+  [configuration file](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/configuration/file-configuration.md), e.g.
+  
+  ```yaml
+  sensitiveData:
+      - attribute: url.query
+        redaction: REDACT_ALL_URL_VALUES
+      - attribute: com.example.customer.name
+        redaction: DROP
+      - attribute: com.example.customer.email
+        redaction: 's/^[^@]*@/REDACTED@/'
+  ```
+
+The above is just an example for how such a configuration may look like, different
+formats are discussed below.
+
+#### Redaction requirement profiles
+
+An end user should be able to pick from a set of redaction requirement profiles
+that cover common configurations:
+
+- `NONE` or `NEVER`: No redaction is applied. This is useful if an end user
+  chooses to run a collector nearby the application that will take care of
+  redaction out of process. If this profile is selected the SDK must make sure
+  that (almost) no overhead remains from the redaction methods.
+- `DEFAULT`: A set of rules that especially make sure that no security-sensitive
```

Comment:
I'm not sure I see a lot of difference between DEFAULT/STRICT/STRICTER.  At a minimum, I see too many "vague" notions where we could wind up bikeshedding what belong in one vs. another.

Can we be more principled in terms of defining:

1. What each profile means.
2. A set of annotations / meta-semantics for attributes?


E.g.

```
NONE - no redaction
NO_USER - redact any information that is input by a user of the system or could be used to identify a user of the system (email, name, IP, etc.)
NO_LOCATION - redact any location information (geo lat/lon, address, etc.)
```

Basically - I don't think the current set is going to be something we can specify.  This needs more rigor and strict lines.

Reply:
Wow, I like that, this is a great idea!